Fix double execution of `Target` expressions in null conditionals

SeeminglyScience opened the pull request:
Fixes #24716

<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

Null conditional expressions (both index and member access) currently evaluate the "target" expression twice. This change saves the "target" result to a temporary variable to compare against.

This change is similar to #12667 which was done to fix null coalesce expressions.

## Disassembly

As compiler changes can be challenging to review, I've included some disassembly of compiled scriptblocks before and after the changes. The module uses reflection to get the delegate compiled internally for the end block, and retrieves the associated LINQ expression tree. That expression tree is then interpreted into representative C#

These are manually formatted and with some of the irrelevant parts removed, full disassembly can be viewed [here](https://gist.github.com/SeeminglyScience/f04355a92f622945f76e55bfd7a253ff).

### Member Expressions

<details>
<summary>

#### Before Changes <sup><sub>(click to expand)</sub></sup>
</summary>

```powershell
$i = 0
$null = ($i++)?.ToString()
```

```cs
// $i = 0
locals.Item009 = 0;

// $null = ($i++)?.ToString()
LanguagePrimitives.IsNull(
    // This is an Expression.Block, not a delegate.
    {
        object @object = locals.Item009;
        locals.Item009 = Fake.Dynamic<Func<CallSite, object, object>>(PSVariableAssignmentBinder.Get())(
            Fake.Dynamic<Func<CallSite, object, object>>(PSUnaryOperationBinder.Get(ExpressionType.Increment))(
                @object));

        return (@object == null) ? 0 : @object;
    })
    ? null
    : Fake.Dynamic<Func<CallSite, object, object>>(
        PSInvokeMemberBinder.Get(
            "ToString",
            classScope: null,
            new CallInfo(0),
            static: false,
            nonEnumerating: false,
            constraints: null))
        (
            // This is an Expression.Block, not a delegate.
            {
                object @object = locals.Item009;
                locals.Item009 = Fake.Dynamic<Func<CallSite, object, object>>(PSVariableAssignmentBinder.Get())(
                    Fake.Dynamic<Func<CallSite, object, object>>(PSUnaryOperationBinder.Get(ExpressionType.Increment))(
                        @object));

                return (@object == null) ? 0 : @object;
            });
```

</details>

<details>
<summary>

#### After - target not strongly typed <sup><sub>(click to expand)</sub></sup>
</summary>

```powershell
$i = 0
$null = ($i++)?.ToString()
```

```cs
// $i = 0
locals.Item009 = 0;

// $null = ($i++)?.ToString()
object object1 =
    // This is an Expression.Block, not a delegate.
    {
        object object2 = locals.Item009;
        locals.Item009 = Fake.Dynamic<Func<CallSite, object, object>>(PSVariableAssignmentBinder.Get())(
            Fake.Dynamic<Func<CallSite, object, object>>(PSUnaryOperationBinder.Get(ExpressionType.Increment))(
                object2));

        return (object2 == null) ? 0 : object2;
    };

LanguagePrimitives.IsNull(object1)
    ? null
    : Fake.Dynamic<Func<CallSite, object, object>>(
        PSInvokeMemberBinder.Get(
            "ToString",
            classScope: null,
            new CallInfo(0),
            static: false,
            nonEnumerating: false,
            constraints: null))(
                object1);
```
</details>

<details>
<summary>

#### After - target strongly typed as valuetype <sup><sub>(click to expand)</sub></sup>
</summary>
```powershell
[int] $i = 0
$null = ($i++)?.ToString()
```

```cs
// [int] $i = 0
locals.Item009 = 0;
// $null = ($i++)?.ToString()
Fake.Dynamic<Func<CallSite, int, object>>(
    PSInvokeMemberBinder.Get(
        "ToString",
        classScope: null,
        new CallInfo(0),
        static: false,
        nonEnumerating: false,
        constraints: null))(
            // This is an Expression.Block, not a delegate.
            {
                int @int = locals.Item009;
                locals.Item009 = Fake.Dynamic<Func<CallSite, object, int>>(
                    PSConvertBinder.Get(typeof(int)))(
                        Fake.Dynamic<Func<CallSite, int, object>>(PSUnaryOperationBinder.Get(ExpressionType.Increment))(
                            @int));

                return @int;
            });
```
</details>
<details>
<summary>

#### After - target strongly typed as ref type <sup><sub>(click to expand)</sub></sup>
</summary>

```powershell
[string] $i = ''
$null = ($i += 'x')?.Length
```

```cs
// [string] $i = ''
locals.Item009 = "";

// $null = ($i += 'x')?.Length
string @string = locals.Item009 = Fake.Dynamic<Func<CallSite, object, string>>(PSConvertBinder.Get(typeof(string)))(
    Fake.Dynamic<Func<CallSite, string, string, object>>(
        PSBinaryOperationBinder.Get(ExpressionType.Add, ignoreCase: true,scalarCompare: false))(
            locals.Item009,
            "x"));

return LanguagePrimitives.IsNull(@string)
    ? null
    : Fake.Dynamic<Func<CallSite, string, object>>(
        PSGetMemberBinder.Get("Length", type: null, static: false, nonEnumerating: false))(
            @string);
```
</details>

### Index Expressions

<details>
<summary>

#### Before Changes <sup><sub>(click to expand)</sub></sup>
</summary>

```powershell
$i = 0
$null = ($i++)?[0]
```

```cs
// $i = 0
locals.Item009 = 0;

// $null = ($i++)?[0]
LanguagePrimitives.IsNull(
    // This is an Expression.Block, not a delegate.
    {
        object @object = locals.Item009;
        locals.Item009 = Fake.Dynamic<Func<CallSite, object, object>>(PSVariableAssignmentBinder.Get())(
            Fake.Dynamic<Func<CallSite, object, object>>(PSUnaryOperationBinder.Get(ExpressionType.Increment))(
                @object));

        return (@object == null) ? 0 : @object;
    })
    ? null
    : Fake.Dynamic<Func<CallSite, object, int, object>>(
        PSGetIndexBinder.Get(
            argCount: 1,
            constraints: null,
            allowSlicing: true))
        (
            // This is an Expression.Block, not a delegate.
            {
                object @object = locals.Item009;
                locals.Item009 = Fake.Dynamic<Func<CallSite, object, object>>(PSVariableAssignmentBinder.Get())(
                    Fake.Dynamic<Func<CallSite, object, object>>(PSUnaryOperationBinder.Get(ExpressionType.Increment))(
                        @object));

                return (@object == null) ? 0 : @object;
            },
            0);
```
</details>

<details>
<summary>

#### After - target not strongly typed <sup><sub>(click to expand)</sub></sup>
</summary>

```powershell
$i = 0
$null = ($i++)?[0]
```

```cs
// $i = 0
locals.Item009 = 0;

// $null = ($i++)?[0]
object object1 =
    // This is an Expression.Block, not a delegate.
    {
        object object2 = locals.Item009;
        locals.Item009 = Fake.Dynamic<Func<CallSite, object, object>>(PSVariableAssignmentBinder.Get())(
            Fake.Dynamic<Func<CallSite, object, object>>(
                PSUnaryOperationBinder.Get(ExpressionType.Increment))
            (object2));

        return (object2 == null) ? 0 : object2;
    };

LanguagePrimitives.IsNull(object1)
    ? null
    : Fake.Dynamic<Func<CallSite, object, int, object>>(
        PSGetIndexBinder.Get(
            argCount: 1,
            constraints: null,
            allowSlicing: true))
        (
            object1,
            0);
```
</details>

<details>
<summary>

#### After - target strongly typed as valuetype <sup><sub>(click to expand)</sub></sup>
</summary>

```powershell
[int] $i = 0
$null = ($i++)?[0]
```

```cs
// [int] $i = 0
locals.Item009 = 0;

// $null = ($i++)?[0]
Fake.Dynamic<Func<CallSite, int, int, object>>(
    PSGetIndexBinder.Get(argCount: 1, constraints: null, allowSlicing: true))
    (
        // This is an Expression.Block, not a delegate.
        {
            int @int = locals.Item009;
            locals.Item009 = Fake.Dynamic<Func<CallSite, object, int>>(PSConvertBinder.Get(typeof(int)))(Fake.Dynamic<Func<CallSite, int, object>>(PSUnaryOperationBinder.Get(ExpressionType.Increment))(@int));

            return @int;
        },
        0);
```
</details>

<details>
<summary>

#### After - target strongly typed as ref type <sup><sub>(click to expand)</sub></sup>
</summary>

```powershell
[string] $i = ''
$null = ($i += 'x')?[0]
```

```cs
locals.Item009 = "";
string @string = locals.Item009 = Fake.Dynamic<Func<CallSite, object, string>>(
    PSConvertBinder.Get(typeof(string)))
    (
        Fake.Dynamic<Func<CallSite, string, string, object>>(
            PSBinaryOperationBinder.Get(
                ExpressionType.Add,
                ignoreCase: true,
                scalarCompare: false))
            (
                locals.Item009,
                "x"));

LanguagePrimitives.IsNull(@string)
    ? null
    : Fake.Dynamic<Func<CallSite, string, int, object>>(
        PSGetIndexBinder.Get(
            argCount: 1,
            constraints: null,
            allowSlicing: true))
        (
            @string,
            0);
```
</details>

## PR Context

<!-- Provide a little reasoning as to why this Pull Request helps and why you have opened it. -->

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
  - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
  - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
  - [x] None
  - **OR**
  - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.5/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
    - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
  - [x] Not Applicable
  - **OR**
  - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
  - [ ] N/A or can only be tested interactively
  - **OR**
  - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
  - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
  - **OR**
  - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
    - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
      - [ ] Issue filed: <!-- Number/link of that issue here -->
    - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
      - [ ] Issue filed: <!-- Number/link of that issue here -->
    - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
      - [ ] Issue filed: <!-- Number/link of that issue here -->
    - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
      - [ ] Issue filed: <!-- Number/link of that issue here -->

SeeminglyScience committed:
Fix double execution of null conditional targets

Fixes #24716

SeeminglyScience commented:
/azp run

daxian-dbw commented on src/System.Management.Automation/engine/parser/Compiler.cs:
```diff
@@ -6324,11 +6324,35 @@ public object VisitMemberExpression(MemberExpressionAst memberExpressionAst)
-                : DynamicExpression.Dynamic(PSGetDynamicMemberBinder.Get(_memberFunctionType, memberExpressionAst.Static), typeof(object), target, Compile(memberExpressionAst.Member));
-
-            return memberExpressionAst.NullConditional ? GetNullConditionalWrappedExpression(target, memberAccessExpr) : memberAccessExpr;
+            return MaybeWrapInNullConditionalExpression(
```
The way you abstract the code with `MaybeWrapInNullConditionalExpression` feels complex and not very readable.
Take this particular call site as an example, how about abstract it in this way? With this approach, we only need to make minor changes to the existing code
- add `var targetToUse =` section
- replace `target` with `targetToUse` when constructing the member-access expression
- change the call to `GetNullConditionalWrappedExpression` slightly
```c#
var target = CompileExpressionOperand(memberExpressionAst.Expression);

// **** Use a temp variable to represent the target for null-conditional ****
var targetToUse = memberExpressionAst.NullConditional
    ? Expression.Parameter(target.Type)
    : target;

// If the ?. operator is used for null conditional check, add the null conditional expression.
// **** Generate the member-access expression with `targetToUse` ****
var memberNameAst = memberExpressionAst.Member as StringConstantExpressionAst;
Expression memberAccessExpr = memberNameAst != null
    ? DynamicExpression.Dynamic(PSGetMemberBinder.Get(memberNameAst.Value, _memberFunctionType, memberExpressionAst.Static), typeof(object), targetToUse)
    : DynamicExpression.Dynamic(PSGetDynamicMemberBinder.Get(_memberFunctionType, memberExpressionAst.Static), typeof(object), targetToUse, Compile(memberExpressionAst.Member));

return memberExpressionAst.NullConditional
    ? GetNullConditionalWrappedExpression((ParameterExpression)targetToUse, target, memberAccessExpr)
    : memberAccessExpr;
```
And change `GetNullConditionalWrappedExpression` to the following:
```c#
private static Expression GetNullConditionalWrappedExpression(ParameterExpression temp, Expression targetExpr, Expression memberAccessExpression)
{
  return Expression.Block(
      typeof(object),
      [temp],
      [
          Expression.Assign(temp, targetExpr),
          Expression.Condition(
              Expression.Call(
                  CachedReflectionInfo.LanguagePrimitives_IsNull,
                  temp.Cast(typeof(object))),
              ExpressionCache.NullConstant,
              memberAccessExpression),
      ]);
}
```

SeeminglyScience replied:
Definitely agree on the complexity 💯 

That suggestion might work! I'll give it a go when I get a minute. Thanks Dongbo!

daxian-dbw reviewed (commented):
@SeeminglyScience Thanks for fixing this! Sorry for the delay in review.

SeeminglyScience committed:
Address complexity feedback

Now uses a temp variable created in the caller instead of a delegate.

Thanks to @daxian-dbw for the pattern recommendation!

SeeminglyScience commented:
/azp run

SeeminglyScience committed:
Merge branch 'master' into SeeminglyScience/issue24716

SeeminglyScience commented:
/azp run

SeeminglyScience committed:
Merge branch 'master' into SeeminglyScience/issue24716

daxian-dbw commented:
@SeeminglyScience Some tests are failing in CI, can you please take a look?
